OpenCraft MVP - Fake PR for review

antoviaque opened the pull request:
See http://console.opencraft.com/ for a working example, and the README for installation instructions.

### Discussions

Reviewers input needed:

- [X] ~~Pylint configuration~~
  - [X] ~~[Which warnings to ignore](https://github.com/open-craft/opencraft/pull/1#discussion-diff-36737650)~~
  - [X] ~~[Max locals](https://github.com/open-craft/opencraft/pull/1#discussion-diff-36738319) & [Max branches](https://github.com/open-craft/opencraft/pull/1#discussion-diff-36738464)~~

### TODO - Current PR

- [X] ~~Small nits~~
- [X] ~~[Clearing temporary files & directories](https://github.com/open-craft/opencraft/pull/1#discussion-diff-36727282)~~
- [X] ~~Use constants for instance & server model status~~
- [X] ~~Web UI loading indicator~~
- [X] ~~Integration tests~~
  - [X] ~~Add an integration test that fully instantiates a sandbox~~
  - [X] ~~Setup CircleCI to run integration tests~~
  - [X] ~~Make CircleCI builds reliable (they currently timeout)~~
- [x] ~~Display a "Please wait" page while DNS are being set for the first time, initially~~
- [x] ~~Add hostname to `/etc/hosts` to avoid warning _"sudo: unable to resolve host xxxx"_~~
- [x] ~~[Add comment for instance statuses, pointing to server status descriptions](https://github.com/open-craft/opencraft/pull/1#discussion_r37565261)~~
- [x] ~~[Shorter & updated sandbox names](https://github.com/open-craft/opencraft/pull/1#discussion_r36607416)~~

### TODO - Future tasks

- [Improve logging management](https://github.com/open-craft/opencraft/pull/1#discussion_r36728597)
  - [Use custom constants for logging levels](https://github.com/open-craft/opencraft/pull/1#discussion-diff-36729765)
- [Switch to OpenStack CLI instead of class](https://github.com/open-craft/opencraft/pull/1#discussion_r36623788)
  - [Use OS_* variables for OpenStack configuration](https://github.com/open-craft/opencraft/pull/1#discussion-diff-36625380)
  - [Use vanilla Ubuntu image from upstream](https://github.com/open-craft/opencraft/pull/1#discussion-diff-36625198)
  - [Implementation-agnostic public IP retreival](https://github.com/open-craft/opencraft/pull/1#discussion_r36624339)

### Suggestions & ideas

- [Links to PRs in Web UI](https://github.com/open-craft/opencraft/pull/1#discussion_r36713342) & Better PR handling via dedicated model
- [Get the build to fail when the ansible run fails (might also require error states for VMs & retries)](https://github.com/open-craft/opencraft/pull/1#discussion_r36624887)
- OpenCraft logo & "Powered by"/licensing footer & as a favicon
- [Relax regex syntax for configuration variables in PRs](https://github.com/open-craft/opencraft/pull/1#discussion_r37372753)
- [Better status transition management - state machine?](https://github.com/open-craft/opencraft/pull/1#discussion_r36733250)
- [Better approach for reboot detection](https://github.com/open-craft/opencraft/pull/1#discussion_r36733404)
- [Load XBlocks by default](https://github.com/open-craft/opencraft/pull/1#discussion_r36607383)

antoviaque commented on instance/templates/instance/ansible/vars.yml:
```diff
@@ -0,0 +1,62 @@
+edx_platform_version: '{{ instance.commit_id }}'
+edx_ansible_source_repo: 'https://github.com/edx/configuration.git'
+configuration_version: 'master'
+
```
From @bradenmacdonald : 

> Is it possible to specify a list of installed XBlocks via ansible configuration?

Yes, we can define a list of XBlocks to load by default, or any additional requirements for that matter, thanks to a PR from Matjaz actually : ) See https://github.com/edx/configuration/pull/1035

We would need to decide on a list of XBlocks to load, and see how/if we pin down their version (we could also get their latest master for now, but that would prevent having reproducible builds, if code pulled changes depending on the build time).

Alternatively, note that there is the possibility to add additional ansible configuration from the PR description - I had forgotten to mention this in the Sandbox HOWTO, fixed: https://github.com/open-craft/documentation/blob/master/sandboxes_howto.md#additional-ansible-configuration . I haven't tested this much yet though, and not at all with the loading of additional requirements or XBlocks, so let me know how that goes when you get a chance to test it.

antoviaque commented on instance/static/html/instance/index.html:
```diff
@@ -0,0 +1,73 @@
+          <div class="large-10 columns">
+            <ul class="instance-description">
+              <li>Instance URL: <a href="{{ selected.instance.url }}" target="_new">{{ selected.instance.url }}</a></li>
+              <li><a href="{{ selected.instance.github_branch_url }}" target="_new">{{ selected.instance.github_base_url }}</a> | {{ selected.instance.branch_name }}</li>
```
From @bradenmacdonald : 

> On the instance page in the console, it provides the "Instance URL" for the LMS - can you add a link to the Studio URL as well?

antoviaque commented on instance/tasks.py:
```diff
@@ -0,0 +1,77 @@
+                branch_name=pr.branch_name,
+            )
+            instance.name = 'PR#{pr.number}: {pr.title} ({pr.username}) - {i.reference_name}'\
+                            .format(pr=pr, i=instance)
```
From @bradenmacdonald : 

> The Platform name is too long I think - see e.g. http://pr9208.sandbox.opencraft.com/login
> How about just using "OpenCraft Open edX PR NNNN" as the platform name?

It's true that it's a bit long at the moment. Probably the title should be removed yes, and the fork/branch are likely superfluous when there is a PR # already. I like having the commit hash though, it can help being sure which instance we are looking at when we refresh it.

antoviaque replied:
@bradenmacdonald I have added the refresh of the commit hash in the name, and shortened it. I've removed the domain name (which is already in the URL bar of the browser), shortened the fork id (to `org/branch`) and limited the PR title to 4 words, as this tended to be the longest part of the name. It's still a bit long, but I find the information it currently contains useful.

Eg.: "PR#9147: Move problem responses export ... (itsjeyd) - open-craft/OC-791-answer-export (047e2fe)"

dachary commented on instance/openstack.py:
```diff
@@ -0,0 +1,98 @@
+
+import requests
+
+from novaclient.v2.client import Client as NovaClient
```
I highly recommend using the python-openstackclient cli instead of a class. It is well documented and by using -f json it essentially behaves as an API call. Here is an example
<pre>
instance = json.loads(subprocess.check_output("openstack server create ..."))
</pre>

See https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/__init__.py#L576 for a more elaborated example (I've been working on that almost full time during the past few weeks).

antoviaque replied:
@dachary Thanks for the advice - I haven't run into any specific issue with the API class so far, and directly using it allowed to do things like using the auto-retry features of requests & httplib3 (cf https://github.com/open-craft/opencraft/pull/1/files#diff-84b847358091a7dddf73724bcbd2b9d3R57 ), but I can see how the interface of the cli would be more bullet-proof and supported. I'll have a look at switching to it.

dachary commented on instance/openstack.py:
```diff
@@ -0,0 +1,98 @@
+    if not addresses:
+        return None
+
+    # TODO: Ensure it is public
```
I implemented a series of functions that deal with this in an OpenStack implementation agnostic way yesterday. https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/__init__.py#L507 The problem is that some OpenStack implementations (namely OVH) assign public IPs to each instance and do not allow you to handle floating ips. Other OpenStack implementations automatically add a public IP to the instance in addition to the private IP, even if you don't ask for a floating IP. And there are the implementations that require you to explicitly allocate a floating IP if you need one.

The logic I implemented copes with these various situations by

* trying to create a floating IP
* if no floating IP can be created, use **nova list** to retrieve the IP of the instance and hope it is publicly accessible (that works for OVH and another OpenStack cluster that automatically allocates floating IPs)
* if a floating IP can be created, associate it with the instance and use **openstack ip floating list** with the instance id to figure out what is the public IP of the instance. Do not use **nova list** as these floating IP do not necessarily show there.

antoviaque replied:
@dachary Cool, seems like you already untangled that mess, I'll gladly reuse what you did : )

Btw, did you see that OVH added support for floating IPs assignments recently? https://community.runabove.com/share/topic/reversedns-and-multiple-ips#comment-2986 Does that work with your code? The lack of support for it there currently prevents from cleanly replacing VMs after a new one has been provisioned, given the IP change.

dachary replied:
I did not know that, thanks for the tip :-)

dachary replied:
@antoviaque this is a hack that is not available via the OpenStack API though. I can see how they need that to allow their customers to migrate smoothly though.
<pre>
$ openstack ip floating pool list
+---------+
| Name    |
+---------+
| Ext-Net |
+---------+
$ openstack ip floating create Ext-Net
ERROR: openstack The server has either erred or is incapable of performing the requested operation. (HTTP 500) (Request-ID: req-33b3ffcf-76cb-43bf-b825-69f62c6af008)
</pre>

dachary commented on instance/tests/models/factories/server.py:
```diff
@@ -0,0 +1,131 @@
+from instance.tests.models.factories.instance import OpenEdXInstanceFactory
+
+
+# Functions ###################################################################
```
It would be easier and provide better coverage to implement integration tests instead (or in addition). Most of not all the logic of the OpenStack provisioning depends on the OpenStack cluster and there is little value (some value but not much ;-) in functional / unit tests. Here is an example:

https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/test/openstack-integration.py#L113

antoviaque replied:
@dachary The tests using this don't really test the OpenStack provisioning itself, but more the code that deals with the OpenStack API state changes, etc. But I see your point, this wouldn't detect any issues caused by changes coming from the cluster, so it would be good to have a safety net there too. That looks like a significant amount of work to rework it this way though.

antoviaque replied:
@dachary I added an integration test, testing a full provisioning of an edX instance, including VM creation & ansible run. Cf https://github.com/open-craft/opencraft/commit/e095d89f657830a98120b501c2c1e5c3e70e16ef

Since running them require access to a working OpenStack cloud, they are run optionally by `make test`, when the config variables are available. I have setup the CircleCI build to automatically run them on PR builds - it's really neat actually, it does a lot of testing automatically now, with only a few additional lines of code : ) It will likely help preventing the type of regressions I introduced this week, where the test suite was running fine, but the edX instances builds were failing.

Sample build with integration tests: https://circleci.com/gh/open-craft/opencraft/27

TODO: Get the build to fail when the ansible run fails - currently the instance is considered provisioned, regardless of the return value from ansible. This will likely require to also add an error state for VMs, and maybe retries if random ansible failures prove to be an issue.

dachary replied:
great move :-)

chromano replied:
@antoviaque as for the missing items on this one, by "error state for VMs" you mean a new status on the Server model or something different? For retries, does a Retry button below the status icon work?

dachary commented on opencraft/settings.py:
```diff
@@ -0,0 +1,340 @@
+OPENSTACK_REGION = env('OPENSTACK_REGION')
+
+OPENSTACK_SANDBOX_FLAVOR = env.json('OPENSTACK_SANDBOX_FLAVOR', default={"ram": 4096, "disk": 40})
+OPENSTACK_SANDBOX_BASE_IMAGE = env.json('OPENSTACK_SANDBOX_BASE_IMAGE', default={"name": "Ubuntu 12.04"})
```
I discovered that even though every OpenStack implementation provides an Ubuntu 12.04 or Ubuntu 14.04 image, they have been modified to some degree. Since all OpenStack implementation provide glance and very cheap storage, I found it a lot easier to upload the official cloud image and use it instead:

See https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/__init__.py#L42 for an inventory of the locations of cloud images for various operating systems and https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/__init__.py#L108 for a two line function that uploads it to your OpenStack provider.

antoviaque replied:
@dachary Nice, thanks! That's a clean way to get rid of that RunAbove motd header ;p

dachary commented on .env.test:
```diff
@@ -0,0 +1,16 @@
+OPENSTACK_PASSWORD='pass'
+OPENSTACK_TENANT='test-tenant'
+OPENSTACK_AUTH_URL='http://example.com/auth/url'
+OPENSTACK_REGION='test-region'
```
I found it easier and more portable to require that the OS_* variables are in the environment. The problem is that the set of variables and they way they are interpreted varies a lot between OpenStack providers. The **openstack** cli deals with this variance and because it is the standard cli, all providers make sure it just works.

antoviaque replied:
@dachary Good point - and to switch to the client API, setting these would be necessary anyway.

antoviaque commented on instance/static/html/instance/index.html:
```diff
@@ -0,0 +1,73 @@
+            <ul class="instance-description">
+              <li>Instance URL: <a href="{{ selected.instance.url }}" target="_new">{{ selected.instance.url }}</a></li>
+              <li><a href="{{ selected.instance.github_branch_url }}" target="_new">{{ selected.instance.github_base_url }}</a> | {{ selected.instance.branch_name }}</li>
+            </ul>
```
From @e-kolpakov : 

> Show a corresponding PR link somewhere (just for convenience)

Agreed, it would be nice. The issue with adding it currently is that the models don't have any notion of PRs themselves - only the functions that create or update the instance objects know about it, so it's not available to the UI. I'll have another look at this when I work on the review comments, to see if there is a way to add this that isn't too involved.

e-kolpakov commented on README.md:
````diff
@@ -1 +1,147 @@
+To run a single test, use `make test_one`:
+
+```
+$ make testone instance.tests.models.test_server
````
@antoviaque nit: `test_one` or `testone`?

antoviaque replied:
It used to be `testone`, but changed to `test_one` without updating the README - thanks for catching this, updated.

e-kolpakov commented on instance/ansible.py:
```diff
@@ -0,0 +1,107 @@
+
+    result_dict = yaml.load(yaml_str1)
+    for key, value in yaml.load(yaml_str2).items():
+        if key in result_dict and isinstance(result_dict[key], dict) and isinstance(value, dict):
```
@antoviaque nit: it might backfire when/if there are three levels of nested dicts (root, nested1 and nested2). Not sure if it is ever going to happen, but it might make sense to fix now, as it likely would be quite hard to debug when something actually breaks.

antoviaque replied:
@e-kolpakov Good point, will do.

e-kolpakov commented on instance/ansible.py:
```diff
@@ -0,0 +1,107 @@
+    fp = os.fdopen(fd, 'w')
+    fp.write(string)
+    fp.close()
+    # TODO: Delete the temporary file after use
```
@antoviaque good idea. Looks like a job for a context manager.

antoviaque replied:
@e-kolpakov True, it's better to address this now, before the code becomes harder to refactor - the same applies to https://github.com/open-craft/opencraft/pull/1/files#diff-a6e5e2d488262cdeea398bf4f1840b98R37 too.

e-kolpakov commented on instance/gandi.py:
```diff
@@ -0,0 +1,108 @@
+    Gandi API proxy object
+    """
+    def __init__(self):
+        self.client = xmlrpc.client.ServerProxy('https://rpc.gandi.net/xmlrpc/')
```
@antoviaque nit: hardcoded url - maybe it should be a django setting? Or even an `__init__` parameter?

antoviaque replied:
@e-kolpakov It's very tightly coupled to Gandi's proprietary API, that's the only endpoint that would expose it. Although it's true that Gandi might add staging/testing endpoint URLs - will do.

e-kolpakov commented on instance/github.py:
```diff
@@ -0,0 +1,168 @@
+    r = requests.get(url, headers=GH_HEADERS)
+    r.raise_for_status()
+
+    r_pr = r.json()
```
@antoviaque the above three lines are replicated in every function - might make sense to extract a function, e.g.

    def send_request(url):
        ...

e-kolpakov commented on instance/log_exception.py:
```diff
@@ -0,0 +1,42 @@
+    Uses the object logging facilities, ie the following method should be defined:
+    self.log(<log_level_str>, <log_message>)`
+    """
+    def wrapper(self, *args, **kwds): #pylint: disable=missing-docstring
```
@antoviaque nit: `functools.wraps(method)`?

e-kolpakov commented on instance/log_exception.py:
```diff
@@ -0,0 +1,42 @@
+        try:
+            return method(self, *args, **kwds)
+        except:
+            self.log('exception', traceback.format_exc())
```
@antoviaque nit: maybe `logging.CRITICAL`? Haven't got to that `log` definition yet, so if you're already considered it and need to distinguish exception from other critical stuff - let it be. Or, actually, add a constant somewhere close to that `log` method.

antoviaque replied:
@e-kolpakov One reason it can be helpful to differentiate exceptions from other critical messages is that we will need to filter out exception output from the logs displayed to the user - if it happens in the wrong place, it might expose variables end users shouldn't be able to see, like authentication strings. Currently we display these because it's only us getting the logs output, but that would need to change. I haven't completely thought this part through though, so if you have better ideas on how to approach this, don't hesitate.

e-kolpakov replied:
@antoviaque I believe `logging` allows [defining custom logging levels](https://docs.python.org/2/library/logging.html#logging-levels), but never done so myself, and docs are a bit scarse on that. But it might be the right way to do that.

antoviaque replied:
@e-kolpakov I did a bit of digging on this, and it looks like we should be using a custom `logging.Handler`, like for the standard library syslog handler: https://github.com/python/cpython/blob/829b49cbd2e4b1d573470da79ca844b730120f3d/Lib/logging/handlers.py#L680-L769 .

And attaching the instance/server model object to a log record via a custom adapter might also be better suited than the current custom logic of the `log()` methods. https://docs.python.org/3.4/howto/logging-cookbook.html#context-info

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+        """
+        active_server_set = self.active_server_set
+        if not active_server_set:
+            return 'empty'
```
@antoviaque nit: magic string - make constant?

antoviaque replied:
@e-kolpakov Good point - it's also tied the `Server.status` choice field - I'll switch both to use constants.

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+        """
+        Returns the subset of `self.server_set` which aren't terminated
+        """
+        return self.server_set.exclude_terminated()
```
@antoviaque I might be getting something wrong, but looks like `Instance` is more or less backend-agnostic, while `server_set` only exists for `OpenStackServer`. It is the only non-abstract Server model, but if we to add more server types it might make sense to move it into abstract `Server` model, or otherwise make sure this method wont crash and will work as it should.

dachary replied:
I found the cloud APIs difficult to abstract. Their semantic varies in ways that have a structural impact on the code using them. For instance some APIs are blocking, others are not, others are optionally and partially blocking. For instance with OpenStack you can create a volume but you can't make that a blocking call which means you then have to pool the volume status until it is available or fails to be. With OpenStack you can however create a server and make that a blocking operation. But it depends which API you're using (the openstack cli does that but not the native nova cli).

I chose to write code that does not abstract the OpenStack behavior at all so that it is abundantly clear to the maintainer that there is no abstraction intended.

antoviaque replied:
@dachary Yup, I'm afraid of these type of structural incompatibilities. However, I've tried to keep an abstraction layer anyway, because we might need to add support for AWS servers - this is what a lot of clients and community members use, and if we ever get to handle the production instances we currently manage with this, I'm not sure they would all agree to move to OpenStack. Having support for both, even if it is tricky, might help to smoothen the migration path toward OpenStack. I am not sure how realistic it is, but it doesn't cost us much to try to keep an abstraction level in the meantime.

@e-kolpakov For `exclude_terminated` specifically, with the abstractions currently in the code, that ought to be a method common to all `Server` models, you are correct. It's actually a method from the queryset/manager, which doesn't have an abstract class currently though - there are other methods on it, like `create()`, that might require to be moved too; I'll have a look to see what could be done simply.

e-kolpakov replied:
@antoviaque My note was more about `server_set` - it is defined in `OpenStackServer` model, not on base abstract `Server` modelk.

antoviaque replied:
@e-kolpakov Oh, good point - fixed. And `exclude_terminated` is already in the right `ServerQuerySet`. Should be all good now then.

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+    github_organization_name = models.CharField(max_length=50, db_index=True)
+    github_repository_name = models.CharField(max_length=50, db_index=True)
+    github_admin_organization_name = models.CharField(max_length=50, blank=True,
+                                                      default=settings.DEFAULT_ADMIN_ORGANIZATION)
```
@antoviaque: from my experience, 50 chars is not enough for such kinds of data :) 200 is more or less enough. Also, might make sense to use constants for common field lengths.

antoviaque replied:
@e-kolpakov Alright : ) Changed.

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+        else:
+            return []
+
+    def set_to_branch_tip(self, branch_name=None, ref_type=None, commit=True):
```
@antoviaque nit: `commit` is a bit ambiguite here - until the very last line I was thinking that it would be used to fetch a specified commit from branch (despite the name of the function). So it might make sense to give it a better name, e.g. `persist`, `save`, or something.

antoviaque replied:
@e-kolpakov I see your point given that this is dealing with git, but it's kind of a standard attribute name in Django for this type of behavior (see for example https://docs.djangoproject.com/en/1.8/topics/forms/modelforms/#the-save-method ). I'll add a comment about this though.

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+        if commit:
+            self.save()
+
+    def set_fork_name(self, fork_name, commit=True):
```
@antoviaque same here for `commit`

antoviaque replied:
@e-kolpakov Same reply : )

e-kolpakov replied:
@antoviaque no wonder :)

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+            self.branch_name = branch_name
+        if ref_type is not None:
+            self.ref_type = ref_type
+        self.log('info', 'Setting instance {} to tip of branch {}'.format(self, self.branch_name))
```
@antoviaque `info` - should be `logging.INFO`or custom constant, not magic string.

antoviaque replied:
@e-kolpakov I'd rather not use `logging.INFO` for this - the value is stored in the DB, and besides the fact that it would expose a logging interna, that's an integer and would be stored as such in the DB (ie, slimmer, but less explicit). Switching to custom constants works for me though, good point.

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+        """
+        Set the organization and repository based on the GitHub fork name
+        """
+        self.log('info', 'Setting fork name for instance {}: {}'.format(self, fork_name))
```
@antoviaque I'll stop repeating the nit about `'info'` from now on - there are likely tens of such calls, and ideally they all should use a constant.

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+        if self.github_organization_name != fork_org \
+                or self.github_repository_name != fork_repo:
+            self.github_organization_name = fork_org
+            self.github_repository_name = fork_repo
```
@antoviaque nit: flat is better than nested. Consider rewriting to

    if self.github_org_name == fork_org and self.github_repo_name == for_repo:
        return

    # the rest of it

e-kolpakov commented on instance/models/instance.py:
```diff
@@ -0,0 +1,436 @@
+        The ansible inventory (list of servers) as a string
+        """
+        inventory = ['[app]']
+        for server in self.server_set.filter(status='booted').order_by('created'):
```
@antoviaque server statuses should also be constants.

e-kolpakov commented on instance/models/logging_mixin.py:
```diff
@@ -0,0 +1,115 @@
+        instance_logentry = next_instance_logentry()
+        server_logentry = next_server_logentry()
+
+        while True:
```
@antoviaque what happens here is a `merge` step in `mergesort`. SO, the more clear implementation would be:

    instance_logenrty = next_instance_logentry()
    server_logenrty = next_server_logentry()

    while instance_logentry and server_logentry:
        if server_logentry.created < instance_logentry.created:
             # append server entry and advance iterator
        else:
             # append instance entry and advence iterator

     while server_logentry:
           # append server entry and advance iterator
     while instance_logentry:
          # append instance entry and advance iterator

Basically the first while consumes everything until one of the iterators is empty, than the rest of the code just appends anything left in the remaining iterator to the end. As you can see, much less ifs :)

antoviaque replied:
@e-kolpakov Nice change - it definitely makes things clearer than trying to do everything within a single loop, thanks.

I have also moved the definition of `next_instance_logentry()` and `next_server_logentry()` to a generic helper function, as the change made mccabe complain about the method complexity.

e-kolpakov commented on instance/models/logging_utils.py:
```diff
@@ -0,0 +1,37 @@
+    """
+    Get the integer code for a log level string
+    """
+    if level == 'exception':
```
@antoviaque this special use case supports the idea of using `logging.*` constants directly.

e-kolpakov commented on instance/models/server.py:
```diff
@@ -0,0 +1,274 @@
+    ('stopping', 'Stopping - Stopping temporarily'),
+    ('stopped', 'Stopped - Stopped temporarily'),
+    ('terminating', 'Terminating - Stopping forever'),
+    ('terminated', 'Terminated - Stopped forever'),
```
@antoviaque server status keys should be constants

e-kolpakov commented on instance/models/server.py:
```diff
@@ -0,0 +1,274 @@
+        self.log('info', 'Waiting for server {} to reach status {}...'.format(self, target_status))
+        while True:
+            self.update_status()
+            if isinstance(target_status, str):
```
@antoviaque nit: more concise:

    target_statuses = [target_status] if isinstance(target_status, str) else target_status
    while True:
         self.update_status()
         if self.status in target_statuses:
             break
         time.sleep(1)
    return self.status

e-kolpakov commented on instance/models/server.py:
```diff
@@ -0,0 +1,274 @@
+
+        return public_addr['addr']
+
+    def update_status(self, provisioned=False, rebooting=False):
```
@antoviaque nit: should probably raise exceptions for invalid state transitions (i.e. `provisioned=True` and `status != 'booted'`)

antoviaque replied:
@e-kolpakov Yes, I'm not very satisfied with the `update_status` method currently. I thought of adding exceptions in these cases, but the method was starting to look unwieldy complex to read. Currently I've made it the responsibility of the calling method to check for the status change (like in https://github.com/open-craft/opencraft/pull/1/files#diff-c40bbeb7cb644b7f9da62d92619b34e6R246 ), but I don't really like it either. One of the issues is that there are lots of different types of transitions, which are very different conceptually,. And sometimes the changes come from the VM itself (starting up, booting), sometimes from the caller or even the instance level (thus the arguments).

I'll try to think of this some more - if anyone has a suggestion for a better approach to status transition, I gladly take it.

e-kolpakov replied:
@antoviaque looks like a job for [state machine](https://en.wikipedia.org/wiki/Event-driven_finite-state_machine). I could try building one when I'm done with GP v2.

antoviaque replied:
@e-kolpakov I would definitely be glad to look at a PR with this approach - I've never had a chance to play much with state machine, it would be a good opportunity to get educated : )

bradenmacdonald replied:
+1 for state machine :)

e-kolpakov commented on instance/models/server.py:
```diff
@@ -0,0 +1,274 @@
+
+        # TODO: Find a better way to wait for the server shutdown and reboot
+        # Currently, without sleeping here, the status would immediately switch back to ready,
+        # as SSH is still available until the reboot terminates the SSHD process
```
@antoviaque maybe wait for `not is_port_open(self.public_ip, 22)`?

antoviaque replied:
@e-kolpakov It would be a bit better than the current naive sleep, but not that much more. We could also miss the time slot during which the port is closed - VMs can reboot pretty fast. The better approach I was thinking about was to for example compare the VM uptime - but that requires to ssh to the instance, so it was more involved than I had the time for. I tried to look for an OpenStack VM attribute or function that would give this, but I surprisingly didn't find one.

@dachary I imagine this must be a common use case - is there a best practice here?

dachary replied:
I don't think there really is one. Some people recommend using cloud-init **phone home** http://cloudinit.readthedocs.org/en/latest/topics/examples.html#call-a-url-when-finished but that requires that there is a publicly available IP to do that. I chose to have a function dealing with all kind of ssh cnx errors and looking for cloud-init termination string in the log files instead, but I'm not entirely happy about it https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/__init__.py#L167 and the user-data file that reboots during cloud-init https://github.com/dachary/teuthology/blob/wip-6502-openstack-v3/teuthology/openstack/openstack-centos-6.5-user-data.txt

bradenmacdonald replied:
@antoviaque What about sending a command like `bash -c "sleep 5; shutdown -r now" & service ssh stop` that will synchronously stop the SSH server and then asynchronously (but deterministically) reboot the server? (The single ampersand between those commands is intentional.) As soon as that command has completed, you can wait for port 22 to open again and then you know that the server has rebooted.

e-kolpakov commented on instance/static/html/instance/index.html:
```diff
@@ -0,0 +1,73 @@
+    </ul>
+
+    <div class="row" ng-if="!instanceList">
+      <p>...</p>
```
@antoviaque this is totally confusing - first time I got to the instance manager I thought it was broken, because it was loading instances in the background without any visual hint. I believe having something like "Loading instances, please wait..." here would be helpful.

antoviaque replied:
@e-kolpakov Yes, I agree this is confusing. "Loading instances..." is also what was there originally, but that's also shown when `instanceList` is actually loaded, but empty (ie when no instance exist yet). I'll need to separate the two cases, but haven't had time to look at this yet.

e-kolpakov commented on instance/tests/models/test_logging.py:
```diff
@@ -0,0 +1,75 @@
+        now = datetime(2015, 8, 5, 18, 7, tzinfo=pytz.utc)
+        sec = timedelta(seconds=1)
+
+        with patch('django.utils.timezone.now') as mock_now:
```
@antoviaque nit: there's a freezegun lib that allows time-travelling in more "backend-agnostic" fashion (i.e. it patches datetime.now, datetime.today and other "now"-returning methods so that they return what you want). Might want to use it.

antoviaque replied:
@e-kolpakov Ah, thanks! I was looking for that lib, which I couldn't remember the name of. Patching datetime directly isn't that much more work though, but it will make the test clearer to read.

e-kolpakov commented on instance/tests/models/test_server.py:
```diff
@@ -0,0 +1,277 @@
+
+        self.assertEqual(server.status, 'new')
+        server.start()
+        self.assertEqual(mock_create_server.mock_calls, [call(
```
@antoviaque `assert_called_once_with`?

e-kolpakov commented on instance/tests/models/test_server.py:
```diff
@@ -0,0 +1,277 @@
+        server.reboot()
+        self.assertEqual(server.status, 'rebooting')
+        server.os_server.reboot.assert_called_once_with(reboot_type='SOFT')
+        self.assertEqual(mock_sleep.mock_calls, [call(30)])
```
@antoviaque nit: `mock_sleep.assert_called_once_with(30)`?

e-kolpakov commented on instance/tests/test_ansible.py:
```diff
@@ -0,0 +1,120 @@
+        self.yaml_dict2 = {
+            'testb': 'secondb with unicode «ταБЬℓσ2»',
+            'testc': 'secondc',
+            'test_dict': {
```
@antoviaque somewhere at the top I mentioned that `yaml_merge` likely won't work with nested dicts. So:

* If decided that such behavior is correct - add test to excercise it.
* If decided to update so that it recursively merges any depth of nested dicts - add test for it.

e-kolpakov commented on pylintrc:
```diff
@@ -0,0 +1,388 @@
+    locally-disabled,
+    too-few-public-methods,
+    too-many-ancestors,
+    unused-argument,
```
@antoviaque nit: `unused-argument` and `no-init` sometimes provide a useful hints - so I would not disable them globally.

antoviaque replied:
@e-kolpakov I've found them to be much more trouble to deal with than they help; if I'm in the minority I'll remove them from this list, but personally I'd rather not have to deal with these warnings.

e-kolpakov commented on pylintrc:
```diff
@@ -0,0 +1,388 @@
+good-names=i,j,k,ex,Run,_
+
+# Bad variable names which should always be refused, separated by a comma
+bad-names=foo,bar,baz,toto,tutu,tata
```
@antoviaque nit: Why there aren't `titi`, `tete` and `tyty` in thies list? :)

Most likely a default pylint config, so just ignore this note :)

antoviaque replied:
@e-kolpakov Yes, that's the default ;p

e-kolpakov commented on pylintrc:
```diff
@@ -0,0 +1,388 @@
+[DESIGN]
+
+# Maximum number of arguments for function / method
+max-args=10
```
@antoviaque nit: I believe there's a 7±2 rule that basically means normal people can't keep in mind more that 9 (and sometimes 5) items. So to me it might make sense to reduce this to 6 or 7.

antoviaque replied:
@e-kolpakov Sounds good, changed `max-args` to 7.

e-kolpakov commented on pylintrc:
```diff
@@ -0,0 +1,388 @@
+ignored-argument-names=_.*
+
+# Maximum number of locals for function / method body
+max-locals=15
```
@antoviaque uhm, this is a bit higher than I feel comfortable. 10 or 12 maybe?

antoviaque replied:
@e-kolpakov That's the default - pylint is generally pretty conservative; for example for https://github.com/open-craft/opencraft/pull/1/files#diff-d63124266d93a6e1c5d4adcf74238461R71 pylint counts 12 local variables already.

e-kolpakov commented on pylintrc:
```diff
@@ -0,0 +1,388 @@
+max-returns=6
+
+# Maximum number of branch for function / method body
+max-branches=12
```
@antoviaque and this definitely needs to go down to at most 8 or 9. I find tracing through the function with more than 8-9 branches quite tedious, and in general, the less branches the better.

antoviaque replied:
@e-kolpakov Again that's the default, I tried to avoid cheating by setting convenient values in the pylintrc just to pass it. : ) I understand the concern to keep things simple, but getting stricter than pylint and mccabe seem pushing it a bit far, imho.

bradenmacdonald replied:
Since so many of these pylint settings are defaults, can you not just leave them out? Why are they all explicitly included?

antoviaque replied:
@bradenmacdonald This file was generated with `pylint --generate-rcfile`, which contains all the default options, which is handy when we need to edit something, and will provide the original value in the diff when we do changes. For this PR it's true that it's confusing though - I should probably have commited the initial default rc before any other change, and only included the diff with the default in this PR.

antoviaque commented:
@dachary @e-kolpakov Thanks for the reviews, lots of useful comments. I've done a first pass, mostly to answer where needed, and addressed a few minor nits. I'll block more time to address the bigger points.

antoviaque commented:
Note: If you have pulled the code since Thursday, you will want to pull it again to get a fix. I had introduced a bug in https://github.com/open-craft/opencraft/commit/2f76e693837b5196f65bd88ef590778f6168c237 which would make the ansible run fail, due to a temporary file being deleting too early. It is fixed in https://github.com/open-craft/opencraft/commit/c5cb17f2bc304dc5de4265086825dbdd1c6f3227.

antoviaque commented:
@e-kolpakov Besides a couple of remaining open points, where comments are welcomed before I actually rework them, like the logging refactoring or the reboot detection, I believe I have addressed most of your points. If you have any comments on the changes made, or if I missed anything, let me know.

@mtyaka @bradenmacdonald When you get a chance, it would be helpful to have your review, as well as your comments on some of the discussions in the comments, above. I still have a few points to address from @dachary (switch to nova CLI instead of using Python classes, making the public IP retrieval more cross-platform), but it might make sense to start doing some of these improvements in separate PRs to allow others to work on the repo too. But I'll need to get your reviews first for that.

e-kolpakov commented:
@antoviaque I believe I don't have anything yet. Do you need a +1 from me? :) The code is already in master, so I don't think so, but just in case, here you go :+1:

mtyaka commented on README.md:
```diff
@@ -1 +1,156 @@
+
+* User interface: [http://localhost:5000/](http://localhost:2000/)
+* API: [http://localhost:5000/api/](http://localhost:2000/api/)
+* Admin: [http://localhost:5000/admin/](http://localhost:2000/admin/)
```
@antoviaque  The link texts say `localhost:5000`, but the links actually point to `localhost:2000` - they should point to port `5000`.

antoviaque replied:
Good catch - fixed. I have also made the ports/hosts configurable, as you suggested.

mtyaka commented on opencraft/settings.py:
```diff
@@ -0,0 +1,340 @@
+# Django-extensions ###########################################################
+
+SHELL_PLUS = "ipython"
+RUNSERVERPLUS_SERVER_ADDRESS_PORT = '127.0.0.1:5000'
```
@antoviaque Can we wrap this setting in `env(...)`, too? I'm trying to run the server in a vagrant box without port forwarding so accessing it through `127.0.0.1` doesn't work.

mtyaka commented on websocket.py:
```diff
@@ -0,0 +1,37 @@
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "opencraft.settings")
+
+host_port = sys.argv[1] if len(sys.argv) > 1 else 'localhost:2001'
```
@antoviaque Can the host:port be pulled from django settings (and customizable in the `.env` file)? I'm trying to run it in a vagrant box without port forwarding.

mtyaka commented on instance/models/instance.py:
```diff
@@ -0,0 +1,469 @@
+        return super().get(*args, **kwargs)
+
+
+class OpenEdXInstance(AnsibleInstanceMixin, GitHubInstanceMixin, LoggerInstanceMixin, Instance):
```
@antoviaque The model name shows up as `Opend ed x instance` in the django admin. Not at all important, but I guess there must be some way to give it a nice django admin name; this looks like the setting: https://docs.djangoproject.com/en/1.8/ref/models/options/#verbose-name

antoviaque replied:
@mtyaka Ah thanks - I had noticed that too, but I wasn't aware of that setting. Added.

antoviaque commented:
@e-kolpakov Yup, I did need a +1 from you : ) It's in master because there was no code before, but it's still good to complete the formal PR process as usual. Thanks!

antoviaque commented:
@e-kolpakov @mtyaka @bradenmacdonald Since the number of comments above is getting a bit large, I have added a checklist to the PR description. It lists the things left to do, but also the points on which reviewer feedback is still needed. Have a look when you get a chance, it should be hopefully easier to locate these this way : )

bradenmacdonald commented on README.md:
```diff
@@ -1 +1,156 @@
+---------
+
+Create an `.env` file at the root of the repository or set environment variables, customizing the
+settings from `opencraft/settings.py` which are loaded via `env()`.
```
Since many settings have no defaults and must be set in this file, it would be nice to have a working example `.env` development configuration in the readme.

antoviaque replied:
@bradenmacdonald It would be nice, but to give a working `.env` file as an example in the README would require publishing auth credentials, like OpenStack cluster env vars or Gandi API token keys. Without these, you wouldn't really be able to do much - the closest there is to a working `.env` file without real auth credentials is the `.env.test` file, but it only works because tests mock out all the external API calls. Imho for someone setting this up, besides getting such an example from someone who would trust you with these auth tokens, there isn't much of a way around reading the `opencraft/settings.py` file, and going to create accounts on these services to gather the credentials. That's why I included lengthy descriptions of the options in that file, to make this a guided process.

bradenmacdonald commented on instance/models/logging.py:
```diff
@@ -0,0 +1,130 @@
+    """
+    Single log entry for instances
+    """
+    instance = models.ForeignKey(OpenEdXInstance, related_name='logentry_set')
```
Nit: Would be nice to have `class Meta: verbose_name_plural = "Instance Log Entries"` since it currently says "Entrys" in the admin, which is wrong (same thing for server log entry model).

antoviaque replied:
@bradenmacdonald Thanks for the tip for this attribute name, I didn't pay much attention to this - added.

bradenmacdonald commented on instance/static/html/instance/index.html:
```diff
@@ -0,0 +1,77 @@
+<top-bar>
+  <ul class="title-area">
+    <li class="name">
+      <h1><a href="#">OpenCraft - Instances Manager</a></h1>
```
I personally find it confusing that OpenCraft is the name of this instance manager software as well as the company. I would suggest renaming it to "OpenCraft IM" (Instance Manager) or something ?

Also it would be cool to put the OpenCraft logo in the Readme or in a footer as "powered by OpenCraft".

Oh, BTW, I think "Instance**s** Manager" is not wrong but it's not the usual way this would be said in English - when describing a type of manager, the word before is not usually pluralized - e.g. "Project Manager", "Program Manager", "Tool Manager", "Process Manager", or (in this case) "Instance Manager".

antoviaque replied:
@bradenmacdonald The way I think of it is that "OpenCraft" is a suite of tools, including the instance manager, which allows to automate the things we do. So technically that's already the case - the name for the current code would be "OpenCraft Instance Manager". I'll edit the title to remove the dash, and the "s" : ) And +1 for including the logo, yup!

As a side note, this gives me an occasion to explain the larger goal with this code -- to start crystallizing our workflow and ways of working into code. We're pretty unique and good at that, in many ways (decentralized, working openly on free software, focus on quality and contributions, good development & deployment workflow inspired from edX good practices...). 

Other companies & communities could benefit from it. In that context, the reason to call the suite of tools to which the IM belongs "OpenCraft", like the company, is that the two aren't necessarily distinct. Ie it's a way to try to scale some of what we are doing manually, and become more than a development shop - ie focus our brand/name on the product, rather than on the company (like GitHub is called GitHub, not GitHub Repository Manager : ) ). Does that make sense, or am I just full of $%$? : ) Note that's more a long term dream plan than a current reality - for now I'm fine with using OpenCraft IM, since we are currently, as a matter of fact, a development shop. We just don't have to be that forever.

bradenmacdonald commented on README.md:
````diff
@@ -1 +1,156 @@
+
+```
+export PATH="$PATH:$HOME/.local/bin" VIRTUALENVWRAPPER_PYTHON=/usr/bin/python3
+source $HOME/.local/bin/virtualenvwrapper.sh
````
I would suggest adding `workon opencraft` to bashrc as well, or putting it in the example commands below, since it's needed in future logins.

antoviaque replied:
@bradenmacdonald Good point - added to the paragraph below, since it only makes sense to add it to the `.bashrc` when that's the only virtualenv the current user will use. Which isn't the case, for example, for my local dev environment (but might be on yours if you use a VM for that).

bradenmacdonald commented on .env.test:
```diff
@@ -0,0 +1,16 @@
+DEBUG=true
+SECRET_KEY='tests'
+DATABASE_URL='postgresql://postgres@localhost/opencraft'
```
This doesn't work for me:
```
django.db.utils.OperationalError: fe_sendauth: no password supplied
```

I would suggest making this default to `DATABASE_URL='postgresql:///opencraft'`

And add a comment that the developer just has to run the one-time setup command `sudo -u postgres createuser -d opencraft`, where 'opencraft' is replaced with the name of whatever user the app runs under.

antoviaque replied:
@bradenmacdonald Ok, I'll try that. Note that the current approach worked out of the box for me on the stock Ubuntus 14.04 that I tried, as well as CircleCI - have you changed the rights or password of your postgres user? The approach you give should also work there though I think, and it's more generic.

bradenmacdonald commented on README.md:
````diff
@@ -1 +1,156 @@
+```
+
+
+Processus description
````
Do you mean "Processes" or are you intentionally using the French word instead?

antoviaque replied:
@bradenmacdonald I just wasn't aware that this wasn't an English word too :p Now I know! Fixed.

bradenmacdonald commented on instance/ansible.py:
```diff
@@ -0,0 +1,124 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Ansible - Helper dunctions
```
Typo: "dunctions"

bradenmacdonald commented:
Here's a wishlist item for a future PR: When I click on an instance, the URL should change using history.pushState (and support deep linking). Right now the URL is always `http://console.opencraft.com/#/` and it is not possible to do things like send someone else a link to the log or configuration of a particular sandbox.

bradenmacdonald commented on instance/gandi.py:
```diff
@@ -0,0 +1,108 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Gandi DNS - Helper dunctions
```
Another typo "dunctions"

bradenmacdonald commented on instance/api/instance.py:
```diff
@@ -0,0 +1,57 @@
+        Start the (re-)provisioning of an instance
+        """
+        instance = self.get_object()
+        if instance.status not in ('empty', 'ready'):
```
Can these status codes be constants instead of strings?

antoviaque replied:
@bradenmacdonald They should yes - I converted the statuses to constants recently, but I seem to have missed these two. Thanks, fixed.

bradenmacdonald commented on instance/github.py:
````diff
@@ -0,0 +1,164 @@
+    """
+    Extract a settings string from a PR description body
+    """
+    m = re.search("..Settings..\r?\n+```[a-z]*\r?\n((?:.+\r?\n)*)```", pr_body)
````
Ah, I was wondering how to override the ansible settings. Is this documented anywhere?

I think this regex could also be loosened a bit: `..Settings..` won't match `## Settings ##` or `### Settings` (how about `\W+Settings\W*` ?). Also, `\r?\n+` will match "\r\n" and "\n\n\n" but not "\r\n\r\n"

antoviaque replied:
@bradenmacdonald Yes, it is documented at https://github.com/open-craft/documentation/blob/master/sandboxes_howto.md#additional-ansible-configuration . But that's a good idea actually to put it here - especially because that doc is private, as it comes from the previous way of setting up sandboxes. I have added it to the "Configuration" tab in the Web UI.

And I'll see about relaxing the syntax a bit, sounds good.

bradenmacdonald commented on instance/github.py:
```diff
@@ -0,0 +1,164 @@
+
+def get_pr_by_number(fork_name, pr_number):
+    """
+    Returns a PR() namedtuple based on the reponse
```
Nit: `PR()` is not a namedtuple [anymore]. Also, I was confused by this since `PR` is defined at the end of this file. Can you move `PR` so it is defined before it is used?

antoviaque replied:
@bradenmacdonald Fixed the nit. For the class position, I'm planning to turn it into a proper PR model, so that's going to be moved out of this file.

bradenmacdonald commented on instance/github.py:
```diff
@@ -0,0 +1,164 @@
+
+def get_pr_list_from_username(user_name, fork_name):
+    """
+    Retreive the current active PRs for a given user
```
Very minor: "Retrieve" is misspelled as "Retreive" five times.

antoviaque replied:
@bradenmacdonald Thanks - that's one of these words my brain registered with the wrong spelling : ) Fixed throughout the code base.

bradenmacdonald commented on instance/models/instance.py:
```diff
@@ -0,0 +1,476 @@
+    LIVE = 'live'
+    STOPPING = 'stopping'
+    STOPPED = 'stopped'
+    TERMINATING = 'terminating'
```
The only differences between this list and the list of statuses in `Server` are the `EMPTY` and `TERMINATED` statuses. Can you just re-use the server statuses instead of repeating so many constants here?

antoviaque replied:
@bradenmacdonald I hesitated on that - it does make sense to reuse the server statuses constants, but there are two issues with it that make me unsure of the benefits:

1. I would still have to redeclare them here, to be able to use `Instance.EMPTY` for example, without having to know if that's shared from `Server` or specific to `Instance`. Ie I would need to have `EMPTY = 'empty'`, `NEW = Server.NEW`, etc.
2. To have `Server.NEW`, it would require importing the `models.server` module - I'm trying to avoid having model modules importing each other, as this quickly leads to circular imports. I would have to move the constants definition to a 3rd place, and then redefine the class attributes on each...

So not sure if this is worth it? Or do you have a better approach in mind?

bradenmacdonald replied:
You could create a new file status.py which contains Status, InstanceStatus, and ServerStatus definitions / state machines.... but that's perhaps getting complicated.

The minimum I would suggest you change for sure is adding some documentation to the constants above. They're documented in server.py but for someone who doesn't know that, it's very difficult to know things like what the difference between "BOOTED" and "STARTED" is or between "ACTIVE", "READY", and "LIVE" is.

antoviaque replied:
@bradenmacdonald Ok, I have added a comment to this section, to make sure someone reading this is aware that fuller descriptions are available in `server.py`.

bradenmacdonald commented on instance/models/instance.py:
```diff
@@ -0,0 +1,476 @@
+
+# Git #########################################################################
+
+class VersionControlInstanceMixin(models.Model):
```
Do these mixins needs to inherit from `models.Model`? You have quite a diamond situation going on here. I think it would be cleaner for the mixins to have no base class unless it's necessary.

antoviaque replied:
@bradenmacdonald I think django needs it to be able to recognize the fields on the parent classes; when I tried it I got a bunch of:

> django.core.exceptions.FieldError: Cannot resolve keyword 'github_repository_name' into field. Choices are: base_domain, certs_version, created, email, forum_version, id, last_provisioning_started, logentry_set, modified, name, notifier_version, protocol, s3_access_key, s3_bucket_name, s3_secret_access_key, server_set, sub_domain, xqueue_version

bradenmacdonald commented on instance/models/instance.py:
```diff
@@ -0,0 +1,476 @@
+
+# Open edX ####################################################################
+
+class OpenEdXInstanceQuerySet(models.QuerySet):
```
This class is not really specific to OpenEdX. I think you should make `DEFAULT_FORK` an attribute/property of the model, and then rename this so it is a generic manager/queryset that can be used for any type of instance.

antoviaque replied:
@bradenmacdonald Good point - done. I attached it to the `GitHubInstance` model, as it does rely on having a GitHub repository and forks.

bradenmacdonald commented on instance/models/instance.py:
```diff
@@ -0,0 +1,476 @@
+    @log_exception
+    def provision(self):
+        """
+        Run the provisioning sequence of the instance, recreating the servers from scratch
```
Nit: It would be good to document the return value in the docstring.

bradenmacdonald commented on instance/models/logging_mixin.py:
```diff
@@ -0,0 +1,128 @@
+
+# Functions ###################################################################
+
+def get_iterator_next_function(iterator, empty_value=None):
```
I don't believe this method is necessary. [Python3's `next()` builtin accepts a second parameter that does the same thing that this method does](https://docs.python.org/3/library/functions.html#next). So you can replace this with a call to `functools.partial`:

```
>>> values = (val for val in ['hello', 'world'])
>>> nexter = partial(next, values, 'end')
>>> nexter()
'hello'
>>> nexter()
'world'
>>> nexter()
'end'
```

e.g. `get_iterator_next_function(server_logentry_set)` is the same as `partial(next, server_logentry_set, None)`

antoviaque replied:
@bradenmacdonald Neat! Thanks for the tip : )

bradenmacdonald commented on instance/models/logging_mixin.py:
```diff
@@ -0,0 +1,128 @@
+    Logging facilities - Logs stored on the model & shared with the client via websocket
+    """
+    class Meta:
+        abstract = True
```
Again, I think you can just remove the `class Meta` and `models.Model` and just make this mixin a regular class, can't you?

bradenmacdonald commented on instance/models/instance.py:
```diff
@@ -0,0 +1,476 @@
+        """
+        Run a playbook against the instance active servers
+        """
+        with clone_configuration_repo() as configuration_repo_path:
```
If we want to allow other types of instances in the future, then `clone_configuration_repo` should be refactored a little. It's named like a generic function but it's very specific to Open edX and OpenCraft.

antoviaque replied:
@bradenmacdonald Done - as part of refactoring needed for the integration tests (which needed to fetch a different repo & branch), I reworked this section and made it fully generic; it's now `with open_repository(repo_url, ref=branch) as repo: ...`

bradenmacdonald commented on instance/serializers.py:
```diff
@@ -0,0 +1,58 @@
+                  'protocol', 'domain', 'url', 'branch_name', 'commit_id', 'github_organization_name',
+                  'github_organization_name', 'github_base_url', 'github_branch_url', 'log_text',
+                  'repository_url', 'updates_feed', 'vars_str', 'created', 'modified', 'status',
+                  'studio_url')
```
Nit: This is a bit messy; formatting it as follows will be way easier to read and produce much nicer diffs:

```python
fields = (
    'id',
    'api_url',
    'active_server_set',
    ...
)
```

bradenmacdonald commented on instance/static/html/instance/index.html:
```diff
@@ -0,0 +1,77 @@
+                                               ng-click="provision(selected.instance)">
+              Reprovision
+            </button>
+            <img src="/static/img/running.png" ng-switch-default />
```
IMHO it's not very obvious what the "running" image means, based on its appearance. It should also have a tooltip and alt text.

bradenmacdonald commented on instance/static/html/instance/index.html:
```diff
@@ -0,0 +1,77 @@
+          <pre class="instance-log">{{ selected.instance.log_text }}</pre>
+        </tab>
+        <tab heading="Configuration">
+          <pre class="configuration">{{ selected.instance.vars_str }}</pre>
```
This would be a great place to add some in-place documentation that reminds users of what to do if they want to change the configuration.

bradenmacdonald commented on instance/static/js/instance.js:
```diff
@@ -0,0 +1,132 @@
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// App configuration //////////////////////////////////////////////////////////
```
No `use strict` in this file?

bradenmacdonald commented on instance/tests/test_github.py:
```diff
@@ -0,0 +1,170 @@
+    """
+    Test cases for GitHub helper functions & API calls
+    """
+    def test_fork_name2tupe(self):
```
Typo: `tupe`

bradenmacdonald commented on requirements.txt:
```diff
@@ -0,0 +1,89 @@
+djangorestframework==3.2.1
+dodgy==0.1.7
+factory-boy==2.5.2
+freezegun==0.3.5
```
This requirements file is huge! Can we split it up into one file for production requirements, and another for test requirements?

antoviaque replied:
@bradenmacdonald Yup, there are a lot of dependencies on this project. The advantage of the current approach is that when you add or update a dependencies, you can do `pip freeze > requirements.txt` and get the diff from git. It's useful to ensure that the virtual env is still in sync, sometimes. On the other side, I'm not sure what there is to gain to split out the test requirements? It's not like if the additional requirements were taking too much space or time to install?

bradenmacdonald replied:
@antoviaque That makes sense. I had no particular reason I guess, just a preference for installing fewer packages where possible, since complex dependency graphs tend to get hard to manage.

bradenmacdonald commented on templates/base.html:
```diff
@@ -0,0 +1,72 @@
+
+    {# Angular #}
+    {% block angular_js %}
+    <script src="//ajax.googleapis.com/ajax/libs/angularjs/1.4.3/angular.js"></script>
```
Nit: you should use `angular.min.js` which is available in the same location, and includes a `.map` file so it should be no worse for debugging.

bradenmacdonald commented:
@antoviaque Took a pass over this and made a bunch of suggestions. Didn't see any significant problems that you don't already know about.

antoviaque commented:
> Here's a wishlist item for a future PR: When I click on an instance, the URL should change using history.pushState (and support deep linking). Right now the URL is always http://console.opencraft.com/#/ and it is not possible to do things like send someone else a link to the log or configuration of a particular sandbox.

+1. The angular router can actually handle this automatically, I just haven't looked at how it works yet. Definitely a good wishlist item for the future.

antoviaque commented:
@bradenmacdonald Thanks for the code review - started addressing your comments above - will keep going through the rest tomorrow.

antoviaque commented:
@bradenmacdonald Finished reviewing your comments - I either addressed them or added the points to the TODO in the PR description. Thanks again!

antoviaque commented:
@e-kolpakov Addressed your comment from yesterday for what is displayed by default while the DNS are updating, initially. This is actually outside of this code since it's just the default DNS wildcard - once floating IPs are supported by the OpenStack providers, we'll be able to handle this more finely, but for now I just replaced the default site under opencraft.com. See http://sdfsdfs.sandbox.opencraft.com/ for an example.

antoviaque commented:
@Kelketek Addressed the issue with the hostname missing in `/etc/hosts` that you mentioned yesterday. The next provisionings should now set it there and prevent the sudo warning you were seeing.

mtyaka commented on .env.test:
```diff
@@ -0,0 +1,16 @@
+DEBUG=true
+SECRET_KEY='tests'
+DATABASE_URL='postgresql:///opencraft'
```
Can we use a different name for the test database, `opencraft-test` maybe? I'd prefer to reserve the name `opencraft` for my development database.

antoviaque replied:
@mtyaka Django already automatically does this, cf https://docs.djangoproject.com/en/1.8/topics/testing/overview/#the-test-database : 

> By default the test databases get their names by prepending test_ to the value of the NAME settings for the databases defined in DATABASES.

But it doesn't hurt to use a different name here, so I'm changing the DB name to `opencraft-test` here.

mtyaka commented on README.md:
```diff
@@ -1 +1,163 @@
+Migrations
+----------
+
+Similarly, migrations are run automatically on startup - for both the development and production
```
I'm not sure I like that... Is this a common setup in python/django world? It makes it very easy to mess up your database if you accidentally check out/deploy a wrong branch. Ideally it would print a warning to let you know you have pending migrations, but not run them automatically.

antoviaque replied:
@mtyaka It's not a common practice, and I see your point. I'm changing it to check for pending migrations on startup instead, and exit with an error if any are found.

mtyaka commented:
Is the [favicon](https://github.com/open-craft/opencraft/blob/master/static/img/favicon/favicon-192x192.png) generic or was it designed for OC Instance Manager?

mtyaka commented on instance/ansible.py:
```diff
@@ -0,0 +1,124 @@
+    result_dict = dict1.copy()
+    for key in dict2:
+        if key in result_dict and isinstance(result_dict[key], dict) and isinstance(dict2[key], dict):
+            result_dict[key] = dict_merge(result_dict[key], dict2[key])
```
Nit: The above would read nicer to me if all instances of `result_dict` were replaced with `dict1` (except the instance where you assign to it, of course).

mtyaka commented on instance/serializers.py:
```diff
@@ -0,0 +1,85 @@
+            'domain',
+            'email',
+            'github_base_url',
+            'github_branch_url', 'log_text',
```
Unusual positiong/formatting of `log_text`.

mtyaka commented:
:+1:

bradenmacdonald commented:
:+1:

antoviaque commented:
@mtyaka The favicon is a generic "copyleft" icon - actually, there is an entry in the TODO to add the OpenCraft logo to the UI. It would make sense to also use it on the favicon, I've added a note about this there.

antoviaque commented:
@mtyaka @bradenmacdonald I've finished addressing some of your pending comments. Thanks again for the reviews, as well as @dachary and @e-kolpakov .

I've also updated the TODO in the PR description: https://github.com/open-craft/opencraft/pull/1#issue-99922395 The list was getting a bit unwiedly and this PR is pretty big already, so I've differentiated between the things to address in the current PR and the larger changes to do in upcoming tasks/PRs. This will make it easier to review additional changes, and allow to move further developments to other branches than master. I've entered the important changes (such as the OpenStack CLI conversion and the logging refactoring) as tasks in the backlog, and the remaining suggestions/ideas in the UserVoice: 

https://opencraft.uservoice.com/forums/280754-opencraft/category/134703-opencraft-im

Votes and comments welcomed.
